cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1823,10 +1451,79 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                            // downgrades which failed in the middle and continue with them.
+                            if (kafkaVersionAnnotation != null) {
+                                if (highestKafkaVersion == null)  {
+                                    highestKafkaVersion = kafkaVersionAnnotation;
+                                } else if (compareDottedVersions(highestKafkaVersion, kafkaVersionAnnotation) < 0) {
+                                    highestKafkaVersion = kafkaVersionAnnotation;
+                                }
+
+                                if (lowestKafkaVersion == null)  {
+                                    lowestKafkaVersion = kafkaVersionAnnotation;
+                                } else if (compareDottedVersions(lowestKafkaVersion, kafkaVersionAnnotation) > 0) {
+                                    lowestKafkaVersion = kafkaVersionAnnotation;
+                                }
+                            }
+
+                            // We find the highest used log.message.format.version. This is later used to validate
+                            // upgrades or downgrades.
+                            if (logMessageFormatAnnotation != null) {
+                                if (highestLogMessageFormatVersion == null)  {
+                                    highestLogMessageFormatVersion = logMessageFormatAnnotation;
+                                } else if (compareDottedVersions(highestLogMessageFormatVersion, logMessageFormatAnnotation) < 0) {
+                                    highestLogMessageFormatVersion = logMessageFormatAnnotation;
+                                }
+                            }
+
+                            // We find the highest used inter.broker.protocol.version. This is later used to validate
+                            // upgrades or downgrades.
+                            if (interBrokerProtocolAnnotation != null)  {
+                                if (highestInterBrokerProtocolVersion == null)  {
+                                    highestInterBrokerProtocolVersion = interBrokerProtocolAnnotation;
+                                } else if (compareDottedVersions(highestInterBrokerProtocolVersion, interBrokerProtocolAnnotation) < 0) {
+                                    highestInterBrokerProtocolVersion = interBrokerProtocolAnnotation;
+                                }
+                            }
+                        }
+
+                        // We decide what is the current Kafka version used and create the KafkaVersionChange object
+                        // describing the situation.
+                        if (lowestKafkaVersion == null)   {
+                            // No version found in Pods or StatefulSet => New cluster or cluster with deleted
+                            // StatefulSet. We treat it as if current version and desired are the same which results in
+                            // noop.
```

Comment:
```suggestion
                            // a noop reconciliation followed by a 2nd reconciliation for actual upgrade/downgrade
```

Reply:
Can you explain the new cluster case mentioned?

Reply:
You mean here or in the comment? 

New cluster means that you just created the Kafka CR and this is the first reconciliation of the cluster => so everything is created for the first time and no upgrade or downgrade matters. In this case the StatefulSet does not exist yet. This is a legitimate noop. There is no version change.

Other case would be that the StatefulSet was deleted either manually or in some disaster situation. In that case all we can do is roll out the new STS based on the CR. We do not know what version was used before it was deleted, but in any case there is nothing to roll. So again, the noop and rolling out the version as per the Kafka CR is ok I think.

The third use case - this is the one which we can filter out and possibly fail the reconciliation - would be if the StatefulSet exists but does not have the annotation. The version annotation is used since ages. So this would mean some super old version before you added the initial upgrade support. My thoughts were that if someone really tries this, they cannot really expect to have a smooth update so it just rolls through it. But as I said, we can also easily fail this.

Reply:
Thanks for the explanation. I think for the 3rd case failing the reconciliation might be best. We could tell the user to add the annotation themselves?

Reply:
Ok, I updated it accordingly.